[Perf] Embeddings: Use router's O(1) lookup and shared sessions

AlexsanderHamir opened the pull request:
## Title

[Perf] Embeddings: Use router's O(1) lookup and shared sessions

## Relevant issues

<!-- e.g. "Fixes #000" -->

## Pre-Submission checklist

**Please complete all items before asking a LiteLLM maintainer to review your PR**

- [ ] I have Added testing in the [`tests/litellm/`](https://github.com/BerriAI/litellm/tree/main/tests/litellm) directory, **Adding at least 1 test is a hard requirement** - [see details](https://docs.litellm.ai/docs/extras/contributing_code)
- [x] I have added a screenshot of my new test passing locally 
- [x] My PR passes all unit tests on [`make test-unit`](https://docs.litellm.ai/docs/extras/contributing_code)
- [x] My PR's scope is as isolated as possible, it only solves 1 specific problem


## Type

<!-- Select the type of Pull Request -->
<!-- Keep only the necessary ones -->

⚡️Performance Improvement

## Changes

- Use router deployment O(1) lookups.
- Allow ProxyBaseLLMRequestProcessing to accept the aembedding route.
- Route embeddings requests through base_process_llm_request, sharing logging, hook execution, retries, header handling, and shared session logic.

## Performance Gains

<p><strong>Before</strong> | 4 CPUs | 8 GB RAM | 4 Instances | DB | No Redis</p>

Type | Name | # Requests | Median (ms) | 95%ile (ms) | 99%ile (ms) | Average (ms) | Current RPS
-- | -- | -- | -- | -- | -- | -- | --
POST | /embeddings | 229947 | 290 | 5700 | 7200 | 844.22 | 1216.7
Custom | LiteLLM Overhead Duration (ms) | 229941 | 14 | 98 | 130 | 31.89 | 1216.7
  | Aggregated | 459888 | 160 | 1200 | 6800 | 438.06 | 2433.4


<!-- notionvc: a42066a5-404d-48ad-9855-3eb173ecba2e -->

<p><strong>After</strong>  | 4 CPUs | 8 GB RAM | 4 Instances | DB | No Redis</p>

Type | Name | # Requests | Median (ms) | 95%ile (ms) | 99%ile (ms) | Average (ms) | Current RPS
-- | -- | -- | -- | -- | -- | -- | --
POST | /embeddings | 344132 | 230 | 430 | 780 | 262.52 | 995.6
Custom | LiteLLM Overhead Duration (ms) | 344132 | 6 | 28 | 47 | 9.46 | 995.6
  | Aggregated | 688264 | 150 | 360 | 570 | 135.99 | 1991.2


<!-- notionvc: 68b49527-d874-4977-9edf-2cb49444c055 -->


## Manual Test

This test was performed with a real API key using the model `text-embedding-3-large`

<img width="918" height="587" alt="Screenshot 2025-11-07 at 4 52 14 PM" src="https://github.com/user-attachments/assets/3cffb5f6-300a-451b-b06e-24d72410c409" />
<img width="1362" height="499" alt="Screenshot 2025-11-07 at 4 54 16 PM" src="https://github.com/user-attachments/assets/956fca1e-26f8-43b7-828f-85320683a497" />
<img width="1361" height="576" alt="Screenshot 2025-11-07 at 4 54 38 PM" src="https://github.com/user-attachments/assets/cf3e3384-db4b-4fc4-a650-532e4efacfd0" />

AlexsanderHamir committed:
Refactor proxy embeddings to use shared processor

- allow ProxyBaseLLMRequestProcessing to accept the aembedding route so embeddings requests reuse the base pipeline hooks

- route embeddings requests through base_process_llm_request, sharing logging, hook execution, retries, and header handling with chat/responses

- tighten token array decoding logic by using router deployment lookups and the unified error handler

AlexsanderHamir committed:
Fix: Correctly process embedding requests with token arrays

The `test_embedding_input_array_of_tokens` test was failing due to a regression that caused embedding requests with token arrays to be processed incorrectly. This prevented the `aembedding` function from being called as expected.

This was caused by a combination of three distinct issues:

1.  In `litellm/proxy/common_request_processing.py`, the `function_setup` utility was called with `aembedding` as the `original_function` for embedding routes. This has been corrected to `embedding` to ensure proper request setup.

2.  In `litellm/proxy/proxy_server.py`, a `TypeError` occurred because the `get_deployment` method was called with the `model_name` keyword argument instead of the expected `model_id`. This has been corrected. Additionally, the check for token arrays was improved to validate that all elements in the input subarray are integers.

3.  In `litellm/proxy/litellm_pre_call_utils.py`, the check for the `enforced_params` enterprise feature was too strict. It blocked valid requests even when the `enforced_params` list was empty. The condition has been adjusted to trigger the check only for non-empty lists.

Finally, the `test_embedding_input_array_of_tokens` assertion was updated to be more robust. The previous `assert_called_once_with` was overly strict, causing failures when unrelated internal parameters were added to the function call. The test now first asserts that `aembedding` is called and then separately verifies the `model` and `input` arguments. This makes the test more resilient to future changes without sacrificing its ability to catch regressions.

AlexsanderHamir committed:
Merge branch 'main' into litellm_embeddings_fix

AlexsanderHamir committed:
Merge remote-tracking branch 'origin/main' into litellm_embeddings_fix

AlexsanderHamir committed:
test: align proxy embedding assertions

Update the embedding proxy test to match the new request pipeline: keep the data the proxy builds, expect the extra control kwargs, let the post-call hook return the actual response, and assert the normalized 'embeddings' hook type. This proves the refactor still forwards metadata and returns the mocked payload.

AlexsanderHamir committed:
Update proxy exception test

The proxy now forwards additional kwargs (request_timeout, litellm_call_id, litellm_logging_obj) to llm_router.aembedding. The test needs to accept these to match the real call signature and keep validating the error path instead of the kwargs list.

ishaan-jaff commented on litellm/proxy/proxy_server.py:
```diff
@@ -5016,167 +5016,75 @@ async def embeddings(  # noqa: PLR0915
             and len(data["input"]) > 0
             and isinstance(data["input"][0], list)
-            and isinstance(data["input"][0][0], int)
+            and all(isinstance(token, int) for token in data["input"][0])
```
why are we changing this ?

ishaan-jaff reviewed (changes requested):
-

AlexsanderHamir committed:
testing: unsure of this change

I don't remember why I changed this, will revert and see if any tests fail since the manual test isn't failing without it.

AlexsanderHamir committed:
fix: remove unrelated change

This change was not related to the embeddings refactor and actually belonged to a different branch.

AlexsanderHamir committed:
Merge: Fix recurrent merge conflic
